Tests/SwiftParserTest/AttributeTests.swift

```diff
@@ -965,6 +965,272 @@ final class AttributeTests: ParserTestCase {
+      @abi(class Sub: Super, Proto where Assoc: OtherProto)
+      class Sub: Super, Proto where Assoc: OtherProto {}
+      """,
+      experimentalFeatures: [.abiAttribute]
+    )
+
+    // Invalid, but diagnosed in ASTGen
+    assertParse(
+      """
+      @abi(associatedtype AssocTy = T)
+      associatedtype AssocTy
+
+      @abi(deinit {})
+      deinit {}
+
+      enum EnumCaseDeclNotParsedAtTopLevel {
+        @abi(case someCase = 42)
+        case someCase
+      }
+
+      @abi(func fn() {})
+      func fn()
+
+      @abi(init() {})
+      init() {}
+
+      @abi(subscript(i: Int) -> Element { get {} set {} })
+      subscript(i: Int) -> Element {}
+
+      @abi(let c1 = 1, c2 = 2)
+      let c1, c2
+
+      @abi(var v1 = 1, v2 = 2)
+      var v1, v2
+
+      @abi(var v3 { get {} set {} })
+      var v3
+      """,
+      experimentalFeatures: [.abiAttribute]
+    )
+
+    // Caught here
+    assertParse(
+      """
+      @abi(class Sub: Super, Proto where Assoc: OtherProto 1️⃣{})
+      class Sub: Super, Proto where Assoc: OtherProto {}
+
+      @abi(var 2️⃣)
+      var v1
+
+      @abi4️⃣(var v23️⃣
+      var v2
+
+      @abi(5️⃣)
+      func fn2() {}
+
+      @abi6️⃣
+      func fn3() {}
+
+      // Suboptimal:
+      @abi7️⃣ func fn4_abi()8️⃣)
+      func fn4() {}
+      """,
+      diagnostics: [
+        DiagnosticSpec(
+          locationMarker: "1️⃣",
+          message: "class header in '@abi' attribute cannot have a member block",
+          fixIts: ["remove member block"]
+        ),
+        DiagnosticSpec(
+          locationMarker: "2️⃣",
+          message: "expected pattern in variable",
+          fixIts: ["insert pattern"]
+        ),
+        DiagnosticSpec(
+          locationMarker: "3️⃣",
+          message: "expected ')' to end attribute",
+          notes: [
+            NoteSpec(
+              locationMarker: "4️⃣",
+              message: "to match this opening '('"
+            )
+          ],
+          fixIts: ["insert ')'"]
+        ),
+        // FIXME: Custom fix-it with copy of decl it's attached to
+        DiagnosticSpec(
+          locationMarker: "5️⃣",
+          message: "expected argument for '@abi' attribute",
+          fixIts: ["insert attribute argument"]
+        ),
+        DiagnosticSpec(
+          locationMarker: "6️⃣",
+          message: "expected '(', ABI-providing declaration, and ')' in attribute",
+          fixIts: ["insert '(', ABI-providing declaration, and ')'"]
+        ),
+        // FIXME: Suboptimal diagnosis
+        DiagnosticSpec(
+          locationMarker: "7️⃣",
+          message: "expected '(', ABI-providing declaration, and ')' in attribute",
+          fixIts: ["insert '(', ABI-providing declaration, and ')'"]
+        ),
+        DiagnosticSpec(
+          locationMarker: "8️⃣",
+          message: "unexpected code ')' before function"
+        ),
+      ],
+      // FIXME: Would love to have the fix-it delete the space before the close paren on the first line
```

Comment:
You should be able to do this with a logic similar to the following

https://github.com/swiftlang/swift-syntax/blob/0386d87aa047ff4bd8bff4d651463657523485a9/Sources/SwiftParserDiagnostics/DiagnosticExtensions.swift#L95-L100